benchmarks/src/main/java/org/elasticsearch/common/RoundingBenchmark.java

```diff
@@ -0,0 +1,120 @@
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+
+import java.time.ZoneId;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+@Fork(1)
+@Warmup(iterations = 5)
+@Measurement(iterations = 3)
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@State(Scope.Benchmark)
+public class RoundingBenchmark {
+    private static final DateFormatter FORMATTER = DateFormatter.forPattern("date_optional_time");
+
+    @Param({
+        "2000-01-01 to 2020-01-01", // A super long range
+        "2000-10-01 to 2000-11-01", // A whole month which is pretty believable
+        "2000-10-29 to 2000-10-30", // A date right around daylight savings time.
+        "2000-06-01 to 2000-06-02"  // A date fully in one time zone. Should be much faster than above.
+    })
+    public String range;
+
+    @Param({ "java time", "es" })
+    public String rounder;
+
+    @Param({ "UTC", "America/New_York" })
+    public String zone;
+
+    @Param({ "MONTH_OF_YEAR", "HOUR_OF_DAY" })
+    public String timeUnit;
+
+    @Param({ "1", "1000000" })
+    public int count;
+
+    private long min;
+    private long max;
+    private long[] dates;
+    private Supplier<Rounding.Prepared> rounderBuilder;
+
+    @Setup
+    public void buildDates() {
+        String[] r = range.split(" to ");
+        min = FORMATTER.parseMillis(r[0]);
+        max = FORMATTER.parseMillis(r[1]);
+        dates = new long[count];
+        long date = min;
+        long diff = (max - min) / dates.length;
+        for (int i = 0; i < dates.length; i++) {
+            if (date >= max) {
+                throw new IllegalStateException("made a bad date [" + date + "]");
+            }
+            dates[i] = date;
+            date += diff;
+        }
+        Rounding rounding = Rounding.builder(Rounding.DateTimeUnit.valueOf(timeUnit)).timeZone(ZoneId.of(zone)).build();
+        switch (rounder) {
+            case "java time":
+                rounderBuilder = rounding::prepareJavaTime;
+                break;
+            case "es":
+                rounderBuilder = () -> rounding.prepare(min, max);
+                break;
+            default:
+                throw new IllegalArgumentException("Expectd rounder to be [java time] or [es]");
+        }
+    }
+
+    @Benchmark
+    public long round() {
+        long sum = 0;
+        Rounding.Prepared rounder = rounderBuilder.get();
+        for (int i = 0; i < dates.length; i++) {
+            sum += rounder.round(dates[i]);
```

Comment:
It's safer here to sink it into a blackhole instead to [reduce the potential for compiler tricks](https://hg.openjdk.java.net/code-tools/jmh/file/b6f87aa2a687/jmh-samples/src/main/java/org/openjdk/jmh/samples/JMHSample_34_SafeLooping.java).

Reply:
I started with that and found that it added pretty significant overhead, smearing out some of differences. When I removed it the differences became more "sharp". I *think* this sort of thing is a safe way to work around the compiler tricks. `-prof=perfasm` seems to bear that out, but computers are tricky.

Reply:
The thing is that with the current code it might be able to do loop unrolling (but then: I did not have a look at the assembler code yet).

Reply:
My first instinct is that loop unrolling wouldn't bother me too much. I'm not really trying to measure the iteration, just the round calls. If loop unrolling let it optimize the round in fun ways that might make it less valid. This is called in a fairly tight loop in production too. Not this tight. And not one the JVM would be able to unroll though.....

Reply:
I ran the microbenchmarks again once with your implementation and once with a cheaper version of Blackhole (calling a method that is not inlined) and there are cases where the score is vastly different, e.g.:

sum (your approach):

```
Benchmark                                                               (count)                   (range)  (rounder)     (timeUnit)            (zone)  Mode  Cnt            Score            Error   Units
RoundingBenchmark.round                                               100000000  2000-06-01 to 2000-06-02  java time    HOUR_OF_DAY               UTC  avgt   10   5468296772.000 ±   13822182.335   ns/op
RoundingBenchmark.round                                               100000000  2000-06-01 to 2000-06-02         es    HOUR_OF_DAY               UTC  avgt   10    949419080.100 ±    2728670.015   ns/op
RoundingBenchmark.round                                               100000000  2000-06-01 to 2000-06-02  java time    HOUR_OF_DAY  America/New_York  avgt   10  13193144584.600 ±   86152570.014   ns/op
RoundingBenchmark.round                                               100000000  2000-06-01 to 2000-06-02         es    HOUR_OF_DAY  America/New_York  avgt   10    947821052.350 ±    6119635.225   ns/op
```

don't inline:

```
Benchmark                                                               (count)                   (range)  (rounder)     (timeUnit)            (zone)  Mode  Cnt            Score            Error   Units
RoundingBenchmark.round                                               100000000  2000-06-01 to 2000-06-02  java time    HOUR_OF_DAY               UTC  avgt   10   5559418450.700 ±    3268596.659   ns/op
RoundingBenchmark.round                                               100000000  2000-06-01 to 2000-06-02         es    HOUR_OF_DAY               UTC  avgt   10   1363302403.000 ±   16780559.912   ns/op
RoundingBenchmark.round                                               100000000  2000-06-01 to 2000-06-02  java time    HOUR_OF_DAY  America/New_York  avgt   10  13554514668.900 ±   25928733.412   ns/op
RoundingBenchmark.round                                               100000000  2000-06-01 to 2000-06-02         es    HOUR_OF_DAY  America/New_York  avgt   10   1354972403.700 ±   15574272.765   ns/op
```

This means that in some cases your microbenchmark implementation shows numbers that are e.g. in the last case in the table above ~ 1.4 times faster than they are if we don't inline which is more likely in production code. We do see a significant speedup compared to the original implementation in both microbenchmark approaches anyway so I think the actual change makes sense. It's just that the change is probably slightly less pronounced in practice than the microbenchmarks indicate.

Reply:
Blackhole it is then! I'll make the change. I had thought that the synchronization overhead that Blackhole adds was sort of unrealistic. And it might be, but it is more realistic than sum.

Reply:
s/synchronization/volatile/